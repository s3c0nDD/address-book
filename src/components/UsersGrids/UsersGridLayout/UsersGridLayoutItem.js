import React, { useMemo, memo } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import PersonIcon from '@material-ui/icons/Person';
import EmailIcon from '@material-ui/icons/Email';

const useStyles = makeStyles(theme => ({
  root: {
    maxWidth: 260
  },
  media: {
    height: 80,
    [theme.breakpoints.up('sm')]: {
      height: 100,
    },
    [theme.breakpoints.up('md')]: {
      height: 120,
    },
    [theme.breakpoints.up('lg')]: {
      height: 150,
    }
  }
}));

/**
 * Checks props and decide if should not re-render
 * @param {object} prevProps previous props
 * @param {object} nextProps next props
 * @returns {boolean}
 */
const shouldNotRerender = (prevProps, nextProps) => {
  return prevProps.user === nextProps.user;
};

/**
 * User grid layout item component
 * @param {object} props component's props
 * @param {object} props.user user data
 * @param {function} props.onOpenModal user modal open handler
 */
const UsersGridLayoutItem = memo(({ user, onOpenModal }) => {
  const classes = useStyles();

  /**
   * Handle grid layout item media click
   */
  const handleClick = () => {
    onOpenModal(user);
  };

  const fullName = useMemo(() => (
    `${user.name.first} ${user.name.last}`
  ), [user]);

  return (
    <Card className={classes.root}>
      <CardActionArea
        onClick={handleClick}
      >
        <CardMedia
          className={classes.media}
          image={user.picture.large}
          title={`${fullName} face photo`}
        />
        <CardContent>
          <Typography
            gutterBottom
            variant="h5"
            component="h2"
          >
            {fullName}
          </Typography>

          <Grid
            container
            justify="flex-start"
            alignItems="stretch"
          >
            <Grid item xs={3} md={2}>
              <PersonIcon color="primary" />
            </Grid>
            <Grid item xs={9} md={10} zeroMinWidth>
              <Typography
                noWrap
                variant="body2"
                color="textSecondary"
                component="p"
              >
                {user.login.username}
              </Typography>
            </Grid>
          </Grid>

          <Grid container>
            <Grid item xs={3} md={2}>
              <EmailIcon color="primary" />
            </Grid>
            <Grid item xs={9} md={10} zeroMinWidth>
              <Typography
                noWrap
                variant="body2"
                color="textSecondary"
                component="p"
              >
                {user.email}
              </Typography>
            </Grid>
          </Grid>
        </CardContent>
      </CardActionArea>
    </Card>
  );
}, shouldNotRerender);

export default UsersGridLayoutItem;

UsersGridLayoutItem.propTypes = {
  user: PropTypes.shape({
    name: PropTypes.shape({
      first: PropTypes.string.isRequired,
      last: PropTypes.string.isRequired
    }).isRequired,
    picture: PropTypes.shape({
      large: PropTypes.string.isRequired
    }).isRequired,
    login: PropTypes.shape({
      username: PropTypes.string.isRequired
    }).isRequired,
    email: PropTypes.string.isRequired
  }).isRequired,
  onOpenModal: PropTypes.func.isRequired
};