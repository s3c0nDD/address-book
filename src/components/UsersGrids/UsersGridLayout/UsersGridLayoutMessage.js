import React  from 'react';
import PropTypes from 'prop-types';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';

/**
 * User grid layout message component
 * @param {object} props component's props
 * @param {object} props.className classnames object
 * @param {function} props.message text to present
 */
const UsersGridLayoutMessage = ({ className, message }) => {
  return (
    <Grid
      className={className}
      container
      justify="center"
      alignItems="center"
      spacing={4}
    >
      <Grid item>
        <Typography
          variant="h4"
          color="textSecondary"
        >
          {message}
        </Typography>
      </Grid>
    </Grid>
  );
};

export default UsersGridLayoutMessage;

UsersGridLayoutMessage.propTypes = {
  className: PropTypes.any,
  message: PropTypes.string.isRequired
};