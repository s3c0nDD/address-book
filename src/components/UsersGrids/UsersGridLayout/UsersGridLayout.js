import React, { memo } from 'react';
import PropTypes from 'prop-types';

import Grid from '@material-ui/core/Grid';

import UsersGridLayoutItem from './UsersGridLayoutItem';

/**
 * Checks props and decide if should not re-render
 * @param {object} prevProps previous props
 * @param {object} nextProps next props
 * @returns {boolean}
 */
const shouldNotRerender = (prevProps, nextProps) => {
  return prevProps.users === nextProps.users;
};

/**
 * User grid layout component
 * @param {object} props component's props
 * @param {array} props.users users data
 * @param {function} props.onOpenModal user modal open handler
 */
const UsersGridLayout = memo(({ users, onOpenModal }) => {
  return (
    <Grid
      container
      direction="row"
      justify="center"
      alignItems="center"
      spacing={4}
    >
      {users?.map((user) => (
        <Grid
          item
          key={user.id.value}
          xs={6}
          sm={4}
          md={3}
          lg={2}
        >
          <UsersGridLayoutItem
            user={user}
            onOpenModal={onOpenModal}
          />
        </Grid>
      ))}
    </Grid>
  );
}, shouldNotRerender);

export default UsersGridLayout;

UsersGridLayout.propTypes = {
  users: PropTypes.array.isRequired,
  onOpenModal: PropTypes.func.isRequired
};