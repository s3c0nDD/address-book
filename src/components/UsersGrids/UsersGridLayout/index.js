import UsersGridLayout from './UsersGridLayout';
import UsersGridLayoutMessage from './UsersGridLayoutMessage';

export default UsersGridLayout;

export {
  UsersGridLayoutMessage
}