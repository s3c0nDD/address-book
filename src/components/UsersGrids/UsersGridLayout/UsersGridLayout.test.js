import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import userMockData from '../../mocks/userMock';
import UsersGridLayout from './UsersGridLayout'
import UsersGridLayoutItem from './UsersGridLayoutItem'

describe('UsersGridLayout', () => {
  let shallow;
  let wrapper;
  let openModalMock;
  let usersMock = [{ ...userMockData }, { ...userMockData, id: { value: 'foobaz' }}];

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    openModalMock = jest.fn();
    wrapper = shallow(
      <UsersGridLayout
        users={usersMock}
        onOpenModal={openModalMock}
      />
    );
  });

  it('shallowly renders and matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('calls prop onOpenModal for every UsersGridLayoutItem onOpenModal call', () => {
    let layoutItemWrappers = wrapper.find(UsersGridLayoutItem);
    layoutItemWrappers.forEach((node) => {
      node.props().onOpenModal();
    });
    expect(openModalMock).toBeCalledTimes(2);
  });
});