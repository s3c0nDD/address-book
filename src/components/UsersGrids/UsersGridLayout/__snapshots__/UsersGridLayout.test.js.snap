// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersGridLayout shallowly renders and matches snapshot 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  container={true}
  direction="row"
  justify="center"
  spacing={4}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    key="foobar"
    lg={2}
    md={3}
    sm={4}
    xs={6}
  >
    <Memo()
      onOpenModal={[MockFunction]}
      user={
        Object {
          "cell": "111-222-333",
          "email": "johndoe123@google.com",
          "id": Object {
            "value": "foobar",
          },
          "location": Object {
            "city": "Warsaw",
            "country": "Poland",
            "postcode": 56789,
            "state": "Masovian",
            "street": Object {
              "name": "Big Street",
              "number": 98,
            },
          },
          "login": Object {
            "username": "johndoe123",
          },
          "name": Object {
            "first": "John",
            "last": "Doe",
            "title": "Mr",
          },
          "phone": "000-111-222",
          "picture": Object {
            "large": "https://wwww.image.com/data/1",
          },
        }
      }
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
    key="foobaz"
    lg={2}
    md={3}
    sm={4}
    xs={6}
  >
    <Memo()
      onOpenModal={[MockFunction]}
      user={
        Object {
          "cell": "111-222-333",
          "email": "johndoe123@google.com",
          "id": Object {
            "value": "foobaz",
          },
          "location": Object {
            "city": "Warsaw",
            "country": "Poland",
            "postcode": 56789,
            "state": "Masovian",
            "street": Object {
              "name": "Big Street",
              "number": 98,
            },
          },
          "login": Object {
            "username": "johndoe123",
          },
          "name": Object {
            "first": "John",
            "last": "Doe",
            "title": "Mr",
          },
          "phone": "000-111-222",
          "picture": Object {
            "large": "https://wwww.image.com/data/1",
          },
        }
      }
    />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
